sw/device/lib/dif/dif_usbdev.c

```diff
@@ -0,0 +1,934 @@
+                                   USBDEV_USBSTAT_RX_EMPTY);
+
+  return kDifUsbdevResultOK;
+}
+
+dif_usbdev_result_t dif_usbdev_irq_enable(dif_usbdev_t *usbdev,
+                                          dif_usbdev_irq_t irq,
+                                          dif_usbdev_toggle_t state) {
+  if (usbdev == NULL) {
+    return kDifUsbdevResultInvalidArgument;
+  }
+
+  if (!is_valid_irq(irq) || !is_valid_toggle(state)) {
+    return kDifUsbdevResultGeneralError;
+  }
+
+  if (state == kDifUsbdevToggleEnable) {
+    mmio_region_nonatomic_set_bit32(usbdev->base_addr,
+                                    USBDEV_INTR_ENABLE_REG_OFFSET,
+                                    kIrqEnumToBitIndex[irq]);
+  } else {
+    mmio_region_nonatomic_clear_bit32(usbdev->base_addr,
+                                      USBDEV_INTR_ENABLE_REG_OFFSET,
+                                      kIrqEnumToBitIndex[irq]);
+  }
+
+  return kDifUsbdevResultOK;
+}
+
+dif_usbdev_result_t dif_usbdev_irq_get(dif_usbdev_t *usbdev,
+                                       dif_usbdev_irq_t irq, bool *state) {
+  if (usbdev == NULL || state == NULL) {
+    return kDifUsbdevResultInvalidArgument;
+  }
+
+  if (!is_valid_irq(irq)) {
+    return kDifUsbdevResultGeneralError;
+  }
+
+  *state = mmio_region_get_bit32(
+      usbdev->base_addr, USBDEV_INTR_STATE_REG_OFFSET, kIrqEnumToBitIndex[irq]);
+
+  return kDifUsbdevResultOK;
+}
+
+dif_usbdev_result_t dif_usbdev_irq_clear(dif_usbdev_t *usbdev,
+                                         dif_usbdev_irq_t irq) {
+  if (usbdev == NULL) {
+    return kDifUsbdevResultInvalidArgument;
+  }
+
+  if (!is_valid_irq(irq)) {
+    return kDifUsbdevResultGeneralError;
+  }
+
+  mmio_region_write32(usbdev->base_addr, USBDEV_INTR_STATE_REG_OFFSET,
+                      1 << kIrqEnumToBitIndex[irq]);
+
+  return kDifUsbdevResultOK;
+}
+
+dif_usbdev_result_t dif_usbdev_irq_clear_all(dif_usbdev_t *usbdev) {
+  if (usbdev == NULL) {
+    return kDifUsbdevResultInvalidArgument;
+  }
+
+  mmio_region_write32(usbdev->base_addr, USBDEV_INTR_STATE_REG_OFFSET,
+                      UINT32_MAX);
+
+  return kDifUsbdevResultOK;
+}
+
+dif_usbdev_result_t dif_usbdev_irq_disable_all(dif_usbdev_t *usbdev,
+                                               uint32_t *cur_config) {
+  if (usbdev == NULL) {
+    return kDifUsbdevResultInvalidArgument;
+  }
+
+  if (cur_config != NULL) {
+    *cur_config =
+        mmio_region_read32(usbdev->base_addr, USBDEV_INTR_ENABLE_REG_OFFSET);
+  }
+
+  mmio_region_write32(usbdev->base_addr, USBDEV_INTR_ENABLE_REG_OFFSET, 0);
+
+  return kDifUsbdevResultOK;
+}
+
+dif_usbdev_result_t dif_usbdev_irq_restore(dif_usbdev_t *usbdev,
+                                           uint32_t new_config) {
+  if (usbdev == NULL) {
+    return kDifUsbdevResultInvalidArgument;
+  }
+
+  mmio_region_write32(usbdev->base_addr, USBDEV_INTR_ENABLE_REG_OFFSET,
+                      new_config);
+
+  return kDifUsbdevResultOK;
+}
```

Comment:
This is a really interesting design for touching all of the IRQ bits at once. Nice work!

Reply:
Thanks, but the credit goes to @silvestrst :) IIRC, I saw this in the UART DIF library.